toolkit/tools/internal/timestamp_v2/timestamp_v2_mgr.go

```diff
@@ -0,0 +1,620 @@
+//    }
+//    defer timestamp_v2.StampMgr.StopMeasurement()												//stops measuring 'tool/scheduler'
+//
+//
+//
+package timestamp_v2
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/jsonutils"
+)
+
+var (
+	StampMgr *TimeStampManager = nil // A shared TimeInfo object that is by default nil. Library will log an warning if the empty object is called.
+)
+
+const (
+	writeCooldownMilliseconds = 1000 // How often to write data back to file
+)
+
+// A TimeStampManager tracks an active set of timing measurements, allowing nested sub-steps to be added easily. The root of the timing tree
+// is stored in `root`. The most recent node added through StartMeasuringEvent() is recorded in `activeNode`.
+type TimeStampManager struct {
+	root               *TimeStamp // Root of the timing tree
+	activeNode         *TimeStamp // Last automatically positioned node
+	dataFilePath       string     // Path to the output .json file
+	dataFileDescriptor *os.File   // Used to coordinate access to the file via flock
+	lock               sync.Mutex // Mutex for controlling writes to the data from cooperating threads (cooperating threads will share a file descriptor which means they can't use flock)
+	isAtomic           bool       // This manager is operating in exclusive mode, no other manager may edit its file while it exists
+	lastUpdate         time.Time  // When was the data last writtent to file
+	shouldWrite        bool       // Should we actually write to file next time we try
+}
+
+// newStampMgr creates a new instance of the global stamp manager object with a root timing node `rootNode`. The root node may
+//	already contain timing information and nested steps if resuming from disk.
+//
+//	The currently active node will always point to the root node upon creation, regardless of any pre-existing data branching from the root.
+func newStampMgr(rootNode *TimeStamp, outputFilePath string, outputFileDescriptor *os.File, atomicMode bool) (newMgr *TimeStampManager) {
+	newMgr = &TimeStampManager{
+		dataFilePath:       outputFilePath,
+		dataFileDescriptor: outputFileDescriptor,
+		root:               rootNode,
+		isAtomic:           atomicMode,
+		shouldWrite:        true,
+		lastUpdate:         time.Now(),
+	}
+	newMgr.activeNode = newMgr.root
+	return
+}
+
+func ensureNoManager() (err error) {
+	if StampMgr != nil {
+		err = fmt.Errorf("already recording timing data for a tool into file (%s)", StampMgr.dataFilePath)
+		failsafeLoggerWarnf(err.Error())
+	}
+	return
+}
+
+func ensureExistsManager() (err error) {
+	if StampMgr == nil {
+		err = fmt.Errorf("timestamping not initialized yet, can't record data")
+		failsafeLoggerWarnf(err.Error())
+	}
+	return
+}
+
+// establishWriteLocks checks if a write should be performed, and locks the file if so
+func (mgr *TimeStampManager) establishWriteLocks() (err error) {
```

Comment:
We still need some locking here even if we ditch the file lock since the go programs DO run parallel workers in a bunch of places.